rfc-157-edge-compute.md

```diff
@@ -0,0 +1,211 @@
+
+As the only common language shared between the two platforms, JavaScript (or TypeScript) seems like the obvious choice.
+
+The primary benefit TypeScript provides over JavaScript is type safety, including enforcement of `readonly` properties. This will be particularly useful on the Lambda@Edge side because parts of the payload we receive from CloudFront are intended to be read-only. TypeScript enforces these invariants (potentially preventing us from introducing bugs), whereas JavaScript doesn't know how to.
+
+If folks have a strong aversion to TypeScript, JavaScript would also be a good option.
+
+## Architecture
+
+I propose that we build a three-layer architecture, with:
+
+- A platform-agnostic wrapper API to abstract away differences between the platform-specific types (e.g. `Request` and `Response`)
+- Platform-agnostic request and response handlers (corresponding to Lambda@Edge's viewer request and viewer response handlers concepts respectively), where the majority of the code lives, e.g.
+  ```typescript
+  // Takes the request from the client
+  // Returns either a request to be sent to the backend,
+  // or a response to be sent to the client
+  function handle_request(req: Request): Request | Response;
+
+  // Takes the response returned from either the request handler or the backend
+  // Returns the response to be sent to the client
+  function handle_response(resp: Response): Response;
+  ```
+- Platform-specific entrypoints for Compute@Edge and Lambda@Edge, which call into the platform-agnostic code, e.g. (for Compute@Edge)
+  ```typescript
+  function perform_backend_fetch(req: Request): Response; // Provided by Fastly SDK
+
+  function fastly_entrypoint(req: Request): Response {
+    const request_or_response = handle_request(request)
+    const response = (request_or_response instanceof Request) ?
+      perform_backend_fetch(request_or_response) :
+      request_or_response;
+    return handle_response(response);
+  }
+  ```
+
+On the Fastly side, we will also need:
+
+- A VCL service _before_ our Compute@Edge one so that we can continue to use shielding on the `/alerts` route (and potentially roll out shielding for all of GOV.UK), and
+- Another VCL service _after_ our Compute@Edge one so that we can manipulate backend responses before they hit the cache and/or control which backend responses are cacheable.
+
+In other words, we will need a VCL -> Compute@Edge -> VCL "sandwich". Such an approach is explicitly mentioned in the [Fastly docs](https://developer.fastly.com/learning/concepts/service-chaining/#chaining-more-than-two-services) as the only use case where a chain of more than 2 services is recommended.
+
+On the Lambda@Edge side, manipulation of the backend response before it hits the cache could be achieved by building an origin response handler. This would create code duplication between Lambda@Edge and the second VCL service in the "sandwich", but that can't be helped right now.
```

Comment:
I think I need to clarify this paragraph, but what I'm proposing is that the `GOVUK-Account-Session` and `GOVUK-Account-End-Session` handling should happen in a VCL service on the Fastly side, and an origin response handler on the AWS side. This creates duplication because we're doing essentially the same thing in both VCL and TypeScript. When Compute@Edge gains support for controlling caching around backend responses then we can refactor this to remove the duplication.